src/Application/FileProcessor/PhpFileProcessor.php

```diff
@@ -49,24 +49,32 @@ public function process(File $file, Configuration $configuration): void
         }, ApplicationPhase::PARSING());
 
         // 2. change nodes with Rectors
-        $this->refactorNodesWithRectors($file);
-
-        // 3. apply post rectors
-        $this->tryCatchWrapper($file, function (File $file): void {
-            $newStmts = $this->postFileProcessor->traverse($file->getNewStmts());
-
-            // this is needed for new tokens added in "afterTraverse()"
-            $file->changeNewStmts($newStmts);
-        }, ApplicationPhase::POST_RECTORS());
-
-        // 4. print to file or string
-        $this->currentFileProvider->setFile($file);
+        do {
+            $file->changeHasChanged(false);
+            $this->refactorNodesWithRectors($file);
+
+            // 3. apply post rectors
+            $this->tryCatchWrapper($file, function (File $file): void {
+                $newStmts = $this->postFileProcessor->traverse($file->getNewStmts());
+
+                // this is needed for new tokens added in "afterTraverse()"
+                $file->changeNewStmts($newStmts);
+            }, ApplicationPhase::POST_RECTORS());
+
+            // 4. print to file or string
+            $this->currentFileProvider->setFile($file);
+
+            // cannot print file with errors, as print would break everything to original nodes
+            if ($file->hasErrors()) {
+                // cannot print file with errors, as print would b
+                $this->notifyPhase($file, ApplicationPhase::PRINT_SKIP());
+                continue;
+            }
 
-        if ($file->hasErrors()) {
-            // cannot print file with errors, as print would b
-            $this->notifyPhase($file, ApplicationPhase::PRINT_SKIP());
-            return;
-        }
+            $this->tryCatchWrapper($file, function (File $file) use ($configuration): void {
+                $this->printFile($file, $configuration);
+            }, ApplicationPhase::PRINT());
+        } while ($file->hasChanged());
```

Comment:
@TomasVotruba it seems make hanging on tests on packages tests, ref https://github.com/rectorphp/rector/issues/6576 

Add `break` before

```diff 
+      break;
    } while ( $file->hasChanged());
```

seems solve it on symfony-packages tests, but not sure if needs another tweak like check rector has completed the check.

Reply:
Ok, add `break` cause error in current `tests` for multi rector:

```bash
1) Rector\Core\Tests\Issues\DoubleRun\DoubleRunTest::test with data set #0 (Symplify\SmartFileSystem\SmartFileInfo Object (...))
tests/Issues/DoubleRun/Fixture/fixture.php.inc
Failed asserting that string matches format description.
--- Expected
+++ Actual
@@ @@
 
 final class RemoveAll
 {
+    public function __construct()
+    {
+    }
 }
 
 ?>
 ```
 
 so it need a something "status" when the rector should stop running